src/wagtail_vector_index/index/models.py

```diff
@@ -0,0 +1,402 @@
+        )
+        embeddings = Embedding.get_for_instance(object)
+
+        # If the existing embeddings all match on content, we return them
+        # without generating new ones
+        if self._existing_embeddings_match(embeddings, splits):
+            return list(embeddings)
+
+        # Otherwise we delete all the existing embeddings and get new ones
+        embeddings.delete()
+
+        embedding_vectors = embedding_backend.embed(splits)
+        generated_embeddings: MutableSequence[Embedding] = []
+        for idx, returned_embedding in enumerate(embedding_vectors):
+            split = splits[idx]
+            embedding = Embedding.from_instance(object)
+            embedding.vector = returned_embedding
+            embedding.content = split
+            embedding.save()
+            generated_embeddings.append(embedding)
+
+        return generated_embeddings
+
+    def to_documents(
+        self, object: EmbeddableFieldsMixin, *, embedding_backend: BaseEmbeddingBackend
+    ) -> Generator[Document, None, None]:
+        for embedding in self.generate_embeddings(
+            object, embedding_backend=embedding_backend
+        ):
+            yield embedding.to_document()
+
+    def from_document(self, document: Document) -> models.Model:
+        if obj := self.base_model.objects.filter(
+            pk=document.metadata["object_id"],
+            content_type=document.metadata["content_type_id"],
+        ).first():
+            return obj
+        else:
+            raise IndexedTypeFromDocumentError("No object found for document")
+
+    def bulk_to_documents(
+        self,
+        objects: Iterable[EmbeddableFieldsMixin],
+        *,
+        embedding_backend: BaseEmbeddingBackend,
+    ) -> Generator[Document, None, None]:
+        # TODO: Implement a more efficient bulk embedding approach
+        for object in objects:
+            yield from self.to_documents(object, embedding_backend=embedding_backend)
+
+    def bulk_from_documents(
+        self, documents: Iterable[Document]
+    ) -> Generator[models.Model, None, None]:
+        # TODO: Implement a more efficient approach
+        for document in documents:
+            yield self.from_document(document)
+
+
+class ModelVectorIndex(VectorIndex):
+    """A VectorIndex which indexes the results of querysets of EmbeddableFieldsMixin models"""
+
+    querysets: ClassVar[Sequence[models.QuerySet]]
+    converter_class: ClassVar[
+        type[EmbeddableFieldsDocumentConverter]
+    ] = EmbeddableFieldsDocumentConverter
+
+    def _get_querysets(self) -> Sequence[models.QuerySet]:
+        return self.querysets
+
+    def get_converter(self) -> EmbeddableFieldsDocumentConverter:
+        queryset_models = [qs.model for qs in self._get_querysets()]
+        all_the_same = len(set(queryset_models)) == 1
+        if not all_the_same:
+            raise ValueError(
+                "All querysets must be of the same model to use the default converter."
+            )
```

Comment:
I've removed the ability to override `converter_class` now for simplicity. If you want to specify a custom converter you can write your own Index and specify `vector_index_class`.

Maybe this will return in the future

Reply:
`ModelVectorIndex` is now `EmbeddableFieldsVectorIndex`. Maybe `ModelVectorIndex` will return one day.

Reply:
Thanks! That makes it simpler I guess! I do wonder if we should come up with a slightly shorter naming.